### PR TITLE
Pull basic manga/chapter metadata from EPUB files.

### PR DESCRIPTION
**Pull request by Steveice10**
Follow-up to my previous pull request for some additional EPUB enhancements.

Pulls the following information from the first chapter for local manga:
* Author
* Description

Pulls the following information for each chapter:
* Title
* Scanlator (Set to Publisher if found, otherwise Creator)
* Publication Date

Also contains code to fuzzy-match manga titles to remove them from chapter names, for cases in which a manga's title within chapter metadata contains characters that can't be represented in the manga directory name.

### CODE REVIEW

File: app/src/main/java/eu/kanade/tachiyomi/util/storage/EpubFile.kt [hunk 1]
```diff
@@ -44,6 +48,48 @@ class EpubFile(file: File) : Closeable {
...
+        }
+
+        if (publisher != null) {
+            chapter.scanlator = publisher.text()
+        } else if (creator != null) {
+            chapter.scanlator = creator.text()
+        }
+
+        if (date != null) {
+            val parsedDate = ISO8601Utils.parse(date.text(), ParsePosition(0))
```
Comment (arkon) [inline app/src/main/java/eu/kanade/tachiyomi/util/storage/EpubFile.kt:88]: Probably don't want to rely on internal Gson stuff. Can refer to other places in the app that use `SimpleDateFormat`.

File: app/src/main/java/eu/kanade/tachiyomi/source/LocalSource.kt [hunk 2]
```diff
@@ -180,6 +196,40 @@ class LocalSource(private val context: Context) : CatalogueSource {
...
+                }
+
+                if (invalidMangaChar) {
+                    mangaTitleIndex++
+                }
+            } else {
+                chapterNameIndex++
+                mangaTitleIndex++
+            }
+        }
```
Comment (arkon) [inline app/src/main/java/eu/kanade/tachiyomi/source/LocalSource.kt:228]: This seems pretty overkill tbh...

File: app/src/main/java/eu/kanade/tachiyomi/util/storage/EpubFile.kt (hunk 1 above)
Comment (Steveice10) [reply app/src/main/java/eu/kanade/tachiyomi/util/storage/EpubFile.kt:88]: SimpleDateFormat is apparently not ISO 8601 compliant, but I'll try to work out a better way of doing this.

File: app/src/main/java/eu/kanade/tachiyomi/source/LocalSource.kt (hunk 2 above)
Comment (Steveice10) [reply app/src/main/java/eu/kanade/tachiyomi/source/LocalSource.kt:228]: The issue here is that a manga's actual title may contain characters like colons that aren't supported in directory names, and as a result the previous code wouldn't be able to strip the manga title out of chapter titles. The only solution I could think of was to put together code that ignores characters that aren't alphanumeric or whitepsace when trying to remove the manga title.

File: app/src/main/java/eu/kanade/tachiyomi/util/storage/EpubFile.kt [hunk 3]
```diff
@@ -44,6 +49,58 @@ class EpubFile(file: File) : Closeable {
...
+        }
+
+        if (publisher != null) {
+            chapter.scanlator = publisher.text()
+        } else if (creator != null) {
+            chapter.scanlator = creator.text()
+        }
+
+        if (date != null) {
+            val dateFormat = SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ssXXX", Locale.getDefault())
```
Comment (arkon) [inline app/src/main/java/eu/kanade/tachiyomi/util/storage/EpubFile.kt:92]: We can't use `X` since only Android SDK 24+ supports that.

Comment (Steveice10) [reply app/src/main/java/eu/kanade/tachiyomi/util/storage/EpubFile.kt]: Replaced with `ZZZZZ`.

**Pull request merged**